utils.py

```diff
@@ -0,0 +1,487 @@
+import csv
+import glob
+import multiprocessing as mp
+import os
+from typing import Tuple
+import warnings
+
+from absl import logging
+import matplotlib.pyplot as plt
+import numpy as np
+import pandas as pd
+import scipy.io
+import tqdm    
+import xlmhg
+
+import istarmap
+
+
+warnings.simplefilter(action='ignore', category=FutureWarning)
+
+def check_data_type(data_path: str) -> str:
+    """Checks if data is single cell or spatial.
+    
+    If 'filtered_feature_bc_matrix' exists in the expected folder, 
+    data type is considered spatial, and scRNAseq otherwise.
+
+    Args:
+        data_path: path for dataset folder.
+
+    Returns:
+        the data type of the provided dataset, "scRNAseq" or "spatial".
+    """
+
+    counts_path = os.path.join(data_path, "filtered_feature_bc_matrix")
+    is_spatial = os.path.exists(counts_path)
+    if is_spatial:
+        return 'spatial'
+    else:
+        return 'scRNAseq'
+
+
+def detect_data_file(data_path: str) -> str:
+    """Checks scRNAseq file extention and returns detected file name.
+    
+    Supported extensions: txt, tsv or pkl.
+    Prioratizing for pkl file, assuming it contains processed data.
+
+    Args:
+        data_path: path for dataset.
+
+    Return:
+        the detected file name.
+
+    Raises:    
+        Exception if more than one file was detected.
+        Exception if no file of types txt, tsv or pkl was found.
+    """
+
+    pkl_file = glob.glob(data_path + '/*.pkl')
+    txt_files = glob.glob(data_path + '/*.txt')
+    tsv_files = glob.glob(data_path + '/*.tsv')
+    if pkl_file:
```

Comment:
this code repeats and can be extracted into a method. something like:
```
def get_data_file(data_path:str, file_extension:str) -> Optional[str]:
  """..."""
  files = glob.glob(os.path.join(data_path, '*.%s' extension))
  if len(files) > 1:
    raise ...
  return next(iter(files), None)
```

and then you can do:
```
data_file = (
   get_data_file(data_path, 'pkl') or
   get_data_file(data_path, 'txt') or
   get_data_file(data_path, 'tsv')
)
if not data_file:
  raise BlaError(...)

return data_file
```